WIMAR - Onboarding

wimar-odoo opened the pull request.

wimar-odoo committed:
[ADD] estate: Init and manifest

Create empty __init__.py and __manifest__.py with module name, set as application.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1005.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1005)

wimar-odoo committed:
[IMP] estate: Models and Basic Fields

Created the Estate Property model and added it to the __init__.py file. Defined the model’s attributes and fields.

wimar-odoo committed:
[IMP] estate: Security - A Brief Introduction

Created access rights granting full permissions to base.group_user and added them to the __manifest__.py file.

thcl-odoo commented on estate/models/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import estate_property
```
There is a missing "blank" line at the end of the file.
It's a “convention,” but it makes the diff easier to read if there are other lines after it :)

thcl-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,23 @@
+from odoo import fields, models
+
+class EstateProperty(models.Model):
```
```suggestion
from odoo import fields, models


class EstateProperty(models.Model):
```

By convention and for readability, leave two blank lines between the imports and the class

thcl-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,23 @@
+        string="Garden Orientation",
+        selection=[('north', 'North'), ('south', 'South'), ('east', 'East'), ('west', 'West')],
+        help="The orientation of the Garden"
+    )
```
```suggestion
    name = fields.Char(required=True)
    description = fields.Text()
    postcode = fields.Char()
    date_availability = fields.Date(string="Available From")
    expected_price = fields.Float(required=True)
    selling_price = fields.Float()
    bedrooms = fields.Integer()
    living_area = fields.Integer(string="Living Area (sqm)")
    facades = fields.Integer()
    garage = fields.Boolean()
    garden = fields.Boolean()
    garden_area = fields.Integer(string="Total Area (sqm)")
    garden_orientation = fields.Selection(
        selection=[('north', 'North'), ('south', 'South'), ('east', 'East'), ('west', 'West')],
        help="The orientation of the Garden"
    )
```

By default, Odoo generates strings. So if they are the same as the technical name of your field, there is no need to add it :)

thcl-odoo commented on estate/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import models
```
end of file

thcl-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,16 @@
+    'data': [
+        'security/ir.model.access.csv',
+    ],
+}
```
end of file

thcl-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,16 @@
+{
+    'name': "Real Estate",
+    'version': '1.0',
+    'summary'  "Manage properties and buyers."
```
```suggestion
    'summary':  "Manage properties and buyers."
```

there is a missing colon here :D

thcl-odoo commented on .gitignore:
```diff
@@ -127,3 +127,13 @@ dmypy.json
+.run
+*/.run/
+.odev
+*/.odev/
```
also missing line here

thcl-odoo reviewed (commented):
First set of comments :) Overall, it's fine!

Here are a few general comments:
- Try to be consistent with the use of single quotes `'` and `"`. If you start using one of them, be consistent throughout :)
- Be careful with commit conventions (you can check the documentation here: [Documentation](https://www.odoo.com/documentation/19.0/contributing/development/git_guidelines.html#commit-message-structure) )

For example `[ADD] Chapter 1: Init and manifest` => `[ADD] estate: app creation`
`[ADD] Chapter 3: Models and Basic Fields` => `[IMP] estate: add models & basic fields` etc.

wimar-odoo committed:
[IMP] estate: Finally, Some UI to Play With

Created the first action and menu, and added them to the __manifest__.py file.
Linked the action to the menu.

wimar-odoo committed:
[IMP] estate: Basic Views

- Added custom views: List, Form, and Search
- Fixed empty line at EOF to comply with PEP8

wimar-odoo committed:
[FIX] estate: Remove useless string parameter from model attributes

wimar-odoo committed:
[FIX] estate: adjustments after first review

wimar-odoo committed:
[IMP] estate: Chapter 7 - Types, Tags and Offers

- Added property types such as House, Apartment, etc., with corresponding Menus, Actions, and Views
- Added property tags such as Renovated, etc., with corresponding Menus, Actions, and Views
- Added an Offers tab on properties to display all offers related to a specific property

wimar-odoo committed:
[IMP] estate: Chapter 8 - Computed, Inverse, and Onchange Fields

- Added computed, inverse, and onchange fields to improve user convenience and automation:
  total_area, best_price, validity, date_deadline, and garden

wimar-odoo committed:
[IMP] estate: Chapter 9 - Button Actions

- Added buttons to sell or cancel a property, and to accept or refuse an offer

wimar-odoo committed:
[IMP] estate: Chapter 10 - SQL and Python Constraints

- Added SQL constraints to ensure positive prices and unique names
- Added Python constraints to enforce offers at least 90% of the expected price

thcl-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,88 @@
+            if record.state == "cancelled":
+                raise UserError("Cancelled properties cannot be sold.")
+            record.state = "sold"
+        return True
```
It's better to make the error message translatable :)

```py
raise UserError(_("Cancelled properties cannot be sold."))
```

You could also do something like this

```suggestion
		if self.filtered(lambda x: x.state == "cancelled"):
			raise UserError(_("Cancelled properties cannot be sold."))

		self.state = "sold"
		return True
```

thcl-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,88 @@
+        return True
+
+    def action_property_cancel(self):
+        for record in self:
+            if record.state == "sold":
+                raise UserError("Sold properties cannot be cancelled.")
+            record.state = "cancelled"
+        return True
```
same

thcl-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,51 @@
+        selection=[("accepted", "Accepted"), ("refused", "Refused")],
+        copy=False,
+    )
+    partner_id = fields.Many2one("res.partner", string="Buyer",  required=True)
```
linter cries about this space 😅 

```suggestion
    partner_id = fields.Many2one("res.partner", string="Buyer", required=True)
```

thcl-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,51 @@
+        for record in self:
+            record.validity = (record.date_deadline - record.create_date.date()).days
+
+    def action_accept_offer(self):
+        for record in self:
+            if record.property_id.state not in ("accepted", "sold", "cancelled"):
+                record.property_id.state = "accepted"
+                record.status = "accepted"
+                record.property_id.selling_price = self.price
+                record.property_id.buyer_id = self.partner_id
+            else:
+                raise UserError(_("This offer can't be accepted because the property is currently %s.")% record.property_id.state)
+        return True
```
For translations, you should pass variables as a parameter cf : https://www.odoo.com/documentation/19.0/developer/howtos/translations.html#variables


+ Same comment than previous action, example:
```suggestion
    def action_accept_offer(self):
  		if self.filtered(lambda x: x.property_id.state not in ("accepted", "sold", "cancelled")):
  			raise UserError(_("This offer can't be accepted because the property is currently %s.", record.property_id.state))
        
      	self.status = "accepted"
        self.property_id.state = "accepted"
        self.property_id.selling_price = self.price
        self.property_id.buyer_id = self.partner_id               

        return True
```

thcl-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,51 @@
+            if record.status == "accepted":
+                record.property_id.state = "new"
+                record.status = "refused"
+                record.property_id.selling_price = 0.0
+                record.property_id.buyer_id = ""
+        return True
```
```suggestion
    def action_refuse_offer(self):
    	accepted_offers = self.filtered(lambda x: x.status == "accepted")
   
        accepted_offers .property_id.state = "new"
        accepted_offers .status = "refused"
        accepted_offers .property_id.selling_price = 0.0
        accepted_offers .property_id.buyer_id = ""
        
        return True
```

thcl-odoo commented on estate/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+access_estate_property_user,access_estate_property_user,model_estate_property,base.group_user,1,1,1,1
+access_estate_property_type_user,access_estate_property_type_user,model_estate_property_type,base.group_user,1,1,1,1
+access_estate_property_tag_user,access_estate_property_tag_user,model_estate_property_tag,base.group_user,1,1,1,1
+access_estate_property_offer_user,access_estate_property_offer_user,model_estate_property_offer,base.group_user,1,1,1,1
```
missing line break 😄

thcl-odoo reviewed (commented):
New batch of comments but looking good so far :)

wimar-odoo committed:
[IMP] estate: Chapter 11 - Sprinkles

- Automatic and manual ordering
- Decorations and invisible attributes
- Stat button with action
- Domain filters

wimar-odoo committed:
[FIX] estate: PR review adjustments

wimar-odoo committed:
[IMP] estate: Chapter 12 - Inheritance

- Override create method with ondelete constraint
- Inherit res.users to display properties on user form

wimar-odoo committed:
[ADD] estate_account: Chapter 13 - Other modules

- Create module to generate invoices in Invoicing

wimar-odoo committed:
[IMP] estate: Chapter 14 - Kanban view

thcl-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,92 @@
+    _positive_expected_price = models.Constraint(
+        "CHECK(expected_price > 0)",
+        "The expected price must be strictly positive.",
+    )
```
just a small comment like that but nothing mandatory: I feel like it's more readable to have the constraints just under the fields, so they are not "hidden" between all those functions :)

but it's just a personal preference

thcl-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,56 @@
+    @api.model
+    def create(self, vals_list):
+        for vals in vals_list:
+            if vals["price"] <= self.env["estate.property"].browse(vals["property_id"]).best_price:
```
it's better to make the orm method outside of the loop.
For example, you could retrieve every "estate.property" outside of the loop and then filtered them in your loop

thcl-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,56 @@
+        for vals in vals_list:
+            if vals["price"] <= self.env["estate.property"].browse(vals["property_id"]).best_price:
+                raise UserError(_("The offer price must be strictly greater than the current best offer."))
+            self.env["estate.property"].browse(vals["property_id"]).state = "received"
```
same with this line, if you use an intermediate variable with your filtered property

thcl-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,129 @@
+                <field name="living_area" filter_domain="[('living_area', '>', self)]"/>
+                <field name="facades"/>
+                <separator/>
+                <filter string="Available" name="state" domain="['|',('state', '=', 'new'),('state', '=', 'received')]"/>
```
it could be
```suggestion
                <filter string="Available" name="state" domain="[('state', 'in', ['new', 'received'])]"/>
```

thcl-odoo reviewed (commented):
new batch of comments 😄 

+ this commit `[IMP] estate: Chapter 13 - Other modules` could/should be sth like `[ADD] estate_account: Chapter 13 - Other modules`

wimar-odoo committed:
[IMP] awesome_owl: Chapter 1 - Intro Owl Components

wimar-odoo committed:
[FIX] estate: last review adjustments

thcl-odoo commented on awesome_owl/static/src/counter/counter.xml:
```diff
@@ -0,0 +1,13 @@
+
+    <t t-name="awesome_owl.counter">
+        <div>
+            Counter: <t t-esc="state.value"/>
```
```suggestion
            Counter: <t t-out="state.value"/>
```

thcl-odoo commented on awesome_owl/static/src/counter/counter.js:
```diff
@@ -0,0 +1,24 @@
+        this.state.value++;
+        if (this.props.onChange) {
+            this.props.onChange();
+        }
```
just for your information, it's maybe less readable in this case but in JS you can do something like this


```suggestion
        this.props.onChange?.();
```

thcl-odoo commented on awesome_owl/static/src/playground.scss:
```diff
@@ -0,0 +1,3 @@
+.chapter_tuto {
+    border: solid; padding: 10px; border-radius: 5px; margin-bottom: 5px
+}
```
just for lisibility :)

```suggestion
.chapter_tuto {
    border: solid; 
    padding: 10px; 
    border-radius: 5px; 
    margin-bottom: 5px
}
```

thcl-odoo commented on awesome_owl/static/src/playground.xml:
```diff
@@ -2,9 +2,29 @@
+        <div class="chapter_tuto">
+            <Counter onChange.bind="incrementSum" t-props="{'buttonText': 'Custom Incr.'}"/>
+            <Counter onChange.bind="incrementSum"/>
+            <p>The sum id: <t t-esc="state.value"/></p>
```
```suggestion
            <p>The sum id: <t t-out="state.value"/></p>
```

thcl-odoo reviewed (commented):
new batch of comments, wp :p

thcl-odoo commented on awesome_owl/static/src/counter/counter.xml:
```diff
@@ -0,0 +1,13 @@
+            Counter: <t t-esc="state.value"/>
+            <button class="btn btn-primary" t-on-click="increment" style="margin-left: 5px">
+                <t t-out="props.buttonText"/>
+            </button>
```
You can "inline" the t-out like this

```suggestion
            <button class="btn btn-primary" t-on-click="increment" style="margin-left: 5px" t-out="props.buttonText" />
```

thcl-odoo reviewed (commented):
Oups I missed this one

wimar-odoo committed:
[IMP] awesome_dashboard: Chapter 2 - Build a dashboard

thcl-odoo commented on awesome_dashboard/static/src/dashboard/dashboard_configuration/dashboard_configuration.js:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    getItemDescription(item) {
+        return item.description || "Dashboard item";
```
```suggestion
        return item.description || _t("Dashboard item");
```

thcl-odoo commented on awesome_dashboard/static/src/dashboard/dashboard_configuration/dashboard_configuration.xml:
```diff
@@ -0,0 +1,33 @@
+                                t-on-change="() => this.toggleItem(item.backend_attribute)"
+                            />
+                            <label class="form-check-label" t-att-for="'item_' + item.backend_attribute">
+                                <t t-esc="getItemDescription(item)"/>
```
```suggestion
                                <t t-out="getItemDescription(item)"/>
```

thcl-odoo commented on awesome_dashboard/static/src/dashboard/numbercard/numbercard.js:
```diff
@@ -0,0 +1,13 @@
+            type: Number,
+        }
+    }
+}
```
Missing EOL

thcl-odoo commented on awesome_dashboard/static/src/dashboard/dashboard_configuration/dashboard_configuration.xml:
```diff
@@ -0,0 +1,33 @@
+                                class="form-check-input"
+                                t-att-id="'item_' + item.backend_attribute"
+                                t-att-checked="state.selectedItems[item.backend_attribute]"
+                                t-on-change="() => this.toggleItem(item.backend_attribute)"
```
No need to change, but another pattern could have been used like.
Sometimes, it can be useful so you've seen this at least once :)

```suggestion
								t-att-data-backend-attribute="item.backend_attribute"
                                t-on-change="toggleItem"
```

And then update your JS function like this

```js
    toggleItem(event) {
		const backendAttribute = event.currentTarget.getAttribute("backend-attribute")
        this.state.selectedItems[backendAttribute] = !this.state.selectedItems[backendAttribute];
    }
```

thcl-odoo commented on awesome_dashboard/static/src/dashboard/numbercard/numbercard.xml:
```diff
@@ -0,0 +1,11 @@
+<templates xml:space="preserve">
+
+    <t t-name="awesome_dashboard.NumberCard">
+        <p class="text-center"><t t-out="props.title"/></p>
```
could be simplified like this 

```suggestion
        <p class="text-center" t-out="props.title"/>
```

thcl-odoo commented on awesome_dashboard/static/src/dashboard/numbercard/numbercard.xml:
```diff
@@ -0,0 +1,11 @@
+        <p class="text-center"><t t-out="props.title"/></p>
+        <div class="text-center">
+            <t t-out="props.value" class="text-center"/>
+        </div>
```
Same

```suggestion
        <div class="text-center" t-out="props.value" />
```

thcl-odoo commented on awesome_dashboard/static/src/dashboard/piechartcard/piechartcard.js:
```diff
@@ -0,0 +1,15 @@
+            type: Object,
+        },
+    }
+}
```
missing EOL :p

thcl-odoo commented on awesome_dashboard/static/src/dashboard/piechartcard/piechartcard.xml:
```diff
@@ -0,0 +1,9 @@
+        <PieChart data="props.values"/>
+    </t>
+
+</templates>
```
missing eol

thcl-odoo commented on awesome_dashboard/static/src/dashboard/piechartcard/piechartcard.xml:
```diff
@@ -0,0 +1,9 @@
+<templates xml:space="preserve">
+
+    <t t-name="awesome_dashboard.PieChartCard">
+        <p class="text-center"><t t-out="props.title"/></p>
```
```suggestion
        <p class="text-center" t-out="props.title" />
```

thcl-odoo commented on awesome_dashboard/static/src/dashboard/dashboard.js:
```diff
@@ -0,0 +1,93 @@
+                    hiddenItems.push(itemId);
+                }
+            }
+            this.state.hiddenItems = hiddenItems;
```
I'd use an array instead of a dict (it's easier to handle, at list it's more readable) but it's only the tutorial so it's ok.
But this part could be more "modern" like 
```suggestion
            this.state.hiddenItems = Object.entries(config)
			  .filter(([itemId, isVisible]) => !isVisible)
			  .map(([itemId]) => itemId);
```

thcl-odoo commented on awesome_dashboard/static/src/dashboard/dashboard.js:
```diff
@@ -0,0 +1,93 @@
+                visibleItems[itemId] = item;
+            }
+        });
+        return visibleItems;
```
In general, when you want to "migrate" from an array to a single value (here a dict), `.reduce(...)` is the solution :)

(just for your information)

```suggestion
	return Object.entries(this.allItems).reduce((acc, [itemId, item]) => {
	  if (!this.state.hiddenItems.includes(item.backend_attribute)) {
	    acc[itemId] = item;
	  }
	  return acc;
	}, {});
```

thcl-odoo commented on awesome_dashboard/static/src/dashboard/dashboard.js:
```diff
@@ -0,0 +1,93 @@
+            const config = {};
+            for (const item of Object.values(this.allItems)) {
+                config[item.backend_attribute] = !hiddenItems.includes(item.backend_attribute);
+            }
```
Same here for the reduce (but nothing mandatory)

```suggestion
		const config = Object.values(this.allItems).reduce((acc, item) => {
		  acc[item.backend_attribute] = !hiddenItems.includes(item.backend_attribute);
		  return acc;
		}, {});
```

thcl-odoo commented on awesome_dashboard/static/src/dashboard/dashboard.scss:
```diff
@@ -0,0 +1,11 @@
+  margin: 5px;
+  padding: 10px;
+  color: black;
+}
```
missing eol

thcl-odoo commented on awesome_dashboard/static/src/dashboard/dashboard_item.xml:
```diff
@@ -0,0 +1,10 @@
+
+    <t t-name="awesome_dashboard.dashboarditem">
+        <div class="item_card" t-att-style="`width: ${18 * props.size}rem;`">
+            <t t-slot="default"></t>
```
could be autoclosing tag

```suggestion
            <t t-slot="default" />
```

thcl-odoo commented on awesome_dashboard/static/src/dashboard/statistics_service.js:
```diff
@@ -0,0 +1,23 @@
+    },
+};
+
+registry.category("services").add("awesome_dashboard.statistics", statisticsService);
```
missing EOL 😄

wimar-odoo committed:
[FIX] awesome_owl: Last review adjustments

thcl-odoo reviewed (commented):
Good job!

Lastest batch of comments

wimar-odoo committed:
[FIX] awesome_dashboard: Last review adjustments
